src/util/Timer.cpp

```diff
@@ -315,35 +417,24 @@ VirtualClock::crank(bool block)
 }
 
 void
-VirtualClock::postToCurrentCrank(std::function<void()>&& f)
-{
-    asio::post(mIOContext, std::move(f));
-}
-
-void
-VirtualClock::postToNextCrank(std::function<void()>&& f)
+VirtualClock::postToExecutionQueue(std::function<void()>&& f,
+                                   ExecutionCategory&& id)
 {
     std::lock_guard<std::recursive_mutex> lock(mDelayExecutionMutex);
 
     if (!mDelayExecution)
     {
-        // Either we are waiting on io_context().run_one here, or by some
-        // chance run_one was woke up by network activity and postToNextCrank
-        // was called from background thread during of just after that (before
-        // mutex is again taken by crank). In first case we need to post
-        // directly to io_context to wake up run_one(). In second case this
-        // handler will be executed in poll_one(), a bit earlier that we want.
-        // But with current design it would be at most one additional job per
-        // crank.
+        // Either we are waiting on io_context().run_one, or by some chance
+        // run_one was woken up by network activity and postToExecutionQueue was
+        // called from a background thread.
 
-        // One immediate post is enough.
+        // In any case, all we need to do is ensure that we wake up `crank`, we
+        // do this by posting an empty event to the main IO queue
         mDelayExecution = true;
-        asio::post(mIOContext, std::move(f));
-    }
-    else
-    {
-        mDelayedExecutionQueue.emplace_back(std::move(f));
+        asio::post(mIOContext, []() {});
     }
+    mDelayedExecutionQueue.emplace_back(
```

Comment:
Yeah I thought about this but decided not to for now:
in master we have the notion of "post to next crank" to avoid triggering "always worst case" type of situations when we process events.

Also, this may create unfair situations that get somewhat mitigated by performing some IO.

In any case, this seemed to be the more conservative approach.

Reply:
I don't understand this comment. Any callback posted always gets enqueued to mDelayedExecutionQueue and then immediately de-queued and re-enqueued on the appropriate segregated queue before processing the segregated queues. There appears to me that there's no step in which the temporary container is anything other than .. temporary. Just copying into one container and then copying out of that into another container. Why bother?

Reply:
master was interleaving IO between processing of the execution queue, so I preserved that property. If we don't do this, we're pretty much guaranteeing to exceed the timeout limit for processing the execution queue (for example, closing multiple ledgers in one go when catching up)

Reply:
maybe this is the same problem than what I highlighted for not always merging queues

Reply:
Right but .. that timeout limit on the execution queue is there for that reason, precisely to bound the frequency of switching between CPU-bound and IO-bound work. Why not rely on that, and do away with the secondary queue?

Reply:
(Discussed in person -- mandatory interleaving of IO with work generated by each batch of work might or might-not be good in and of itself, so we can leave it for now and experiment later)